DREAM/Optimization/tasdreamOptimizationTests.cpp

```diff
@@ -0,0 +1,203 @@
+    return pass;
+}
+
+// Unit tests for TasOptimization::ParticleSwarm on a single objective function.
+bool testParticleSwarmSingle(ObjectiveFunction f, ParticleSwarmState state, TasDREAM::DreamDomain inside, int iterations,
+                             double optimal_val, double tol) {
+    bool pass = true;
+
+    // Run the particle swarm algorithm.
+    std::minstd_rand park_miller(42);
+    std::uniform_real_distribution<double> unif(0.0, 1.0);
+    auto get_rand = [&]()->double{ return unif(park_miller); };
+    ParticleSwarm(f, iterations, inside, state, 0.5, 2, 2, get_rand);
+
+    // Check optimality and state changes of the run.
+    std::vector<double> best_swarm_point = state.getBestPosition();
+    std::vector<double> best_swarm_value_vec(1);
+    f(best_swarm_point, best_swarm_value_vec);
+    pass = pass and (fabs(best_swarm_value_vec[0] - optimal_val) <= tol);
+    std::vector<bool> init_vector = state.getStateVector();
+    pass = pass and (init_vector[3]);
+
+    // Make sure subsequent runs do not make any strange modifications.
+    ParticleSwarm(f, 1, inside, state, 0.5, 2, 2, get_rand);
+    f(best_swarm_point, best_swarm_value_vec);
+    pass = pass and (fabs(best_swarm_value_vec[0] - optimal_val) <= tol);
+    init_vector = state.getStateVector();
+    pass = pass and (init_vector[3]);
+
+    // TasOptimization::ParticleSwarmState::clearCache().
+    state.clearCache();
+    init_vector = state.getStateVector();
+    pass = pass and (init_vector[0] and init_vector[1] and init_vector[2] and !init_vector[3]);
+
+    return pass;
+}
+
+// Unit tests for TasOptimization::ParticleSwarm on multiple objective functions.
+bool testParticleSwarm(bool verbose, double tol=1e-2) {
+    bool pass = true;
+
+    // l1 norm over the domain [-5, 2] ^ 6.
+    int num_dimensions = 6;
+    int num_particles = 100;
+    int iterations = 200;
+    std::vector<double> lower(num_dimensions, -5);
+    std::vector<double> upper(num_dimensions, 2);
+    TasOptimization::ObjectiveFunctionSingle l1_single =
+            [](const std::vector<double> &x)->double {
+                double sum = 0;
+                for (auto xi : x) sum += fabs(xi);
+                return sum;};
+    TasOptimization::ObjectiveFunction l1 = TasOptimization::makeObjectiveFunction(num_dimensions, l1_single);
+    TasOptimization::ParticleSwarmState state(num_dimensions, num_particles);
+    state.initializeParticlesInsideBox(lower, upper);
+    pass = pass and (testParticleSwarmSingle(l1, state, TasDREAM::hypercube(lower, upper), iterations, 0, tol));
+
+
+    // Six hump-camel function over the domain [-3, 3] x [-2, 2].
+    num_dimensions = 2;
+    num_particles = 50;
+    iterations = 20;
+    lower = {-3.0, -2.0};
+    upper = {3.0, 2.0};
+    TasOptimization::ObjectiveFunctionSingle shc_single =
+            [](const std::vector<double> &x)->double {
+                return (4 - 2.1 * x[0]*x[0] + x[0]*x[0]*x[0]*x[0] / 3) * x[0]*x[0] +
```

Comment:
Use `4.0` and `3.0` in these type of operations, using ints can mess up the result, especially when division is involved.

Reply:
Done.